pkg/gofr/gofr.go

```diff
@@ -203,6 +190,62 @@ func (a *App) Run() {
 
 	wg.Wait()
 }
+func (a *App) setupHTTPServer() {
+	// Add Default routes
+	a.add(http.MethodGet, "/.well-known/health", healthHandler)
+	a.add(http.MethodGet, "/.well-known/alive", liveHandler)
+	a.add(http.MethodGet, "/favicon.ico", faviconHandler)
+
+	if _, err := os.Stat("./static/openapi.json"); err == nil {
+		a.add(http.MethodGet, "/.well-known/openapi.json", OpenAPIHandler)
+		a.add(http.MethodGet, "/.well-known/swagger", SwaggerUIHandler)
+		a.add(http.MethodGet, "/.well-known/{name}", SwaggerUIHandler)
+	}
+
+	a.httpServer.router.PathPrefix("/").Handler(handler{
+		function:  catchAllHandler,
+		container: a.container,
+	})
+
+	var registeredMethods []string
+
+	_ = a.httpServer.router.Walk(func(route *mux.Route, _ *mux.Router, _ []*mux.Route) error {
+		met, _ := route.GetMethods()
+		for _, method := range met {
+			if !contains(registeredMethods, method) { // Check for uniqueness before adding
+				registeredMethods = append(registeredMethods, method)
+			}
+		}
+
+		return nil
+	})
+
+	*a.httpServer.router.RegisteredRoutes = registeredMethods
+}
+
+// Shutdown stops the service(s) and close the application.
+func (a *App) Shutdown(ctx context.Context) (err error) {
+	if a.httpServer != nil {
+		err = errors.Join(err, a.httpServer.Shutdown(ctx))
```

Comment:
those could probably be shutdown simultaneously

Reply:
been thinking something like that

```golang
func (a *App) Shutdown(ctx context.Context) error {
	wg := shutdownGroup{errs: make(chan error)}

	wg.shutdownWhen(ctx,a.httpServer != nil, a.httpServer.Shutdown)
	wg.shutdownWhen(ctx,a.grpcServer != nil, a.grpcServer.Shutdown)
	wg.shutdownWhen(ctx,a.metricServer != nil, a.metricServer.Shutdown)
	wg.shutdownWhen(ctx,a.container != nil, a.container.Close)
	err := wg.WaitAll()

	if err != nil {
		a.container.Logger.Errorf("error while shutting down: %v", err)
		return err
	}

	a.container.Logger.Infof("Application shutdown complete")

	return nil
}

type shutdownGroup struct {
	sync.WaitGroup
	errs chan error
}

func (wg *shutdownGroup) shutdownWhen(ctx context.Context, condition bool, shutdownFunc func(context context.Context) error) {
	if !condition {
		return
	}

	wg.Add(1)

	go func() {
		defer wg.Done()
		wg.errs <- shutdownFunc(ctx)
	}()
}

func (wg *shutdownGroup) WaitAll() (err error) {
	go func() {
		wg.Wait()

		close(wg.errs)
	}()

	for {
		newErr, ok := <-wg.errs
		if !ok {
			return err
		}

		err = errors.Join(err, newErr)
	}
}

```

Reply:
> been thinking something like that
> 
> ```golang
> func (a *App) Shutdown(ctx context.Context) error {
> 	wg := shutdownGroup{errs: make(chan error)}
> 
> 	wg.shutdownWhen(ctx,a.httpServer != nil, a.httpServer.Shutdown)
> 	wg.shutdownWhen(ctx,a.grpcServer != nil, a.grpcServer.Shutdown)
> 	wg.shutdownWhen(ctx,a.metricServer != nil, a.metricServer.Shutdown)
> 	wg.shutdownWhen(ctx,a.container != nil, a.container.Close)
> 	err := wg.WaitAll()
> 
> 	if err != nil {
> 		a.container.Logger.Errorf("error while shutting down: %v", err)
> 		return err
> 	}
> 
> 	a.container.Logger.Infof("Application shutdown complete")
> 
> 	return nil
> }
> 
> type shutdownGroup struct {
> 	sync.WaitGroup
> 	errs chan error
> }
> 
> func (wg *shutdownGroup) shutdownWhen(ctx context.Context, condition bool, shutdownFunc func(context context.Context) error) {
> 	if !condition {
> 		return
> 	}
> 
> 	wg.Add(1)
> 
> 	go func() {
> 		defer wg.Done()
> 		wg.errs <- shutdownFunc(ctx)
> 	}()
> }
> 
> func (wg *shutdownGroup) WaitAll() (err error) {
> 	go func() {
> 		wg.Wait()
> 
> 		close(wg.errs)
> 	}()
> 
> 	for {
> 		newErr, ok := <-wg.errs
> 		if !ok {
> 			return err
> 		}
> 
> 		err = errors.Join(err, newErr)
> 	}
> }
> 
> ```

Your idea is interesting and complete something I had in mind when I suggested adding the context on all closer

- https://github.com/gofr-dev/gofr/pull/781#discussion_r1664620968

What about doing this


```go
func (a *App) Shutdown(ctx context.Context) error {
	wg := shutdownGroup{errs: make(chan error)}

	wg.shutdownWhen(ctx, a.httpServer)
	wg.shutdownWhen(ctx,a.grpcServer)
	wg.shutdownWhen(ctx,a.metricServer) 
	wg.shutdownWhen(ctx,a.container)
	err := wg.WaitAll()

	if err != nil {
		a.container.Logger.Errorf("error while shutting down: %v", err)
		return err
	}

# 	a.container.Logger.Infof("Application shutdown complete")

	return nil
}

type shutdownGroup struct {
	sync.WaitGroup
	errs chan error
}

type Shutdowner interface {
     Shutdown (context.Context) error
}

func (wg *shutdownGroup) shutdownWhen(ctx context.Context, s Shutdowner) {
	if s == nil {
		return
	}

	wg.Add(1)

	go func() {
		defer wg.Done()
		wg.errs <- s.Shutdown(ctx)
	}()
}

func (wg *shutdownGroup) WaitAll() (err error) {
	go func() {
		wg.Wait()

		close(wg.errs)
	}()

	for {
		newErr, ok := <-wg.errs
		if !ok {
			return err
		}

		err = errors.Join(err, newErr)
	}
}
```

Reply:
@ccoVeille Not all of them called `Shutdown` some are called `Close` and somewhere deeper there's even Close without params. 
For example Reader and Writer in Kafka which should be closed before shutdown as it described in kafka-go docs
Would be nice to move shutdowner somewhere else and reuse that in pubsub implementations to (I'm currently working on closing them as well)
![image](https://github.com/gofr-dev/gofr/assets/20914644/95e6f484-37f8-4d2b-83f2-938549d5b193)

Reply:
Also handling context cancellation in method WaitAll would be nice too for finishing when timeout is up

Reply:
OK, it's clearer so yes, maybe you will need to find a name that fit your needs and won't conflict with Shutdown or Close

Reply:
This one is quite obvious

```
type Stopper interface {
   Stop(ctx context.Context) error
}
```
less obvious/easy to miss
```
type Halter interface {
   Halt(ctx context.Context) error
}
```

Reply:
I'll probably do that in separate smaller PR. This one is getting big